feat: custom theme switching

g1eny0ung opened the pull request:
Signed-off-by: Yue Yang <g1enyy0ung@gmail.com>

Resolve #4619.

This PR changes a lot about how to **load**, **set**, **select**, and **sync** the custom themes. I will try to describe it in as much detail as possible. (More on the key steps 😸)

## Definitions

The old `ThemeOptions` will change to `LegacyTheme` and the `Theme` is represented as the current theme.

The only difference between them is that `Theme` must set the mode to `light` and `dark`, while `LegacyTheme` does not.

## Load

To be compatible with the `LegacyTheme` settings, the loading theme will be divided into two parts:

```ts
/**
 * Activate the user preferences.
 *
 * Steps:
 *
 * 1. Load the custom theme.
 *
 * @memberof LSPluginCore
 */
async activateUserPreferences() {
  const { theme: legacyTheme, themes } = this._userPreferences
  const currentTheme = themes[themes.mode]

  // If there is currently a theme that has been set
  if (currentTheme) {
    await this.selectTheme(currentTheme, { effect: false })
  } else if (legacyTheme) {
    // Otherwise compatible with older versions
    await this.selectTheme(legacyTheme, { effect: false })
  }
}
```

The first thing you will look for here is whether there is a `light/dark` theme setting.

## Set

The **set** is also compatible with `LegacyTheme`. This part of the logic will be described in a list below:

- has `LegacyTheme`
  - has mode?
    - Yes -> inject the link and set it to `state` (`(set-state! [:ui/custom-theme (keyword mode)] theme)`) and `localStorage`
    - No -> inject link only
- has `Theme`
  - inject the link and set it, like above

## Select

Then we can **select** the theme on the UI. There are some changes:

- Now, all installed themes will be divided based on the current **mode**, which means you can't select dark themes in light mode.
- The default theme will also be stored if you select it. Its `url` will be `null`, which means the next loading will not inject anymore.
- After selecting, it will save into `.logseq/preferences.json`.

## Sync

The selection is also accompanied by a synchronization operation. The `:ui/custom-theme` in `localStorage` will be synced with the preferences, in every `emit('theme-selected')`.

g1eny0ung commented:
Feel free to pull this request into your local to test. 😆 I will be looking for some feedback.

andelf commented on libs/src/LSPlugin.core.ts:
```diff
@@ -195,17 +209,26 @@ enum PluginLocalLoadStatus {
   LOADED = 'loaded',
   UNLOADED = 'unload',
-  ERROR = 'error'
+  ERROR = 'error',
```
Better avoid code style and whitespace changes.

g1eny0ung replied:
Hmm. We defined `.prettierrc.js`, but the files inside the libs don't follow it.

It looks like we need a git hook that will automatically unify the code before committing. I would like to open a new PR to handle this, with [husky](https://typicode.github.io/husky/#/) and [lint-staged](https://github.com/okonet/lint-staged). What do u think?

andelf replied:
husky did the ci stage thing. personally, I don't think it's very good practice.
Better and code style check to CI.
CC @xyhp915 @logseq-cldwalker

g1eny0ung replied:
The advantage is that developers don't care about any lint checks, and all possible formatting problems can be dealt with locally.

g1eny0ung replied:
I created #4760 to discuss something related to this.

xyhp915 replied:
@g1eny0ung Current lint rules of libs TS files follow StandardJS https://standardjs.com/. There's a prettier package for those rules https://www.npmjs.com/package/prettier-standard. But for sure, will support commit hooks for lint task.

g1eny0ung commented:
Hi @xyhp915. Since the past two weeks have been busy, I've just removed the burden of additional code formatting from the review. PTAL again when you're free. 😃

chrisdlees commented:
Stiiiiiiill looking good, kinda annoying that PRs would be so slow on something so prolific...

g1eny0ung commented:
@tiensonqin @andelf @xyhp915 Mind taking a look when you have time? Hope to get some of your views.

I'll check the changes later to ensure it's still upstream compatible. 😌

tiensonqin commented:
@g1eny0ung Sorry for taking so long! We'll review this PR soon.

g1eny0ung commented:
> @g1eny0ung Sorry for taking so long! We'll review this PR soon.

Never mind. This enhancement isn't very urgent.

xyhp915 commented:
Hi @g1eny0ung I found some installed themes not in the `select` list no matter which mode selected. Could you try this theme plugin?

<img width="409" alt="CleanShot 2022-05-17 at 09 39 01@2x" src="https://user-images.githubusercontent.com/1779837/168710629-89c253f4-b848-4256-9635-a56c3febffd4.png">

And, what do you think about listing all themes in `select` UI? But we could make a small optimization of classifying theme modes as below.

![CleanShot 2022-05-17 at 09 44 35@2x](https://user-images.githubusercontent.com/1779837/168712190-1d55bd89-dc85-4068-9d07-5cf061e71310.png)

g1eny0ung commented:
> Hi @g1eny0ung I found some installed themes not in the `select` list no matter which mode selected. Could you try this theme plugin?

Got it. I'll try it later.

> And, what do you think about listing all themes in `select` UI? But we could make a small optimization of classifying theme modes as below.
> 
> ![CleanShot 2022-05-17 at 09 44 35@2x](https://user-images.githubusercontent.com/1779837/168712190-1d55bd89-dc85-4068-9d07-5cf061e71310.png)

👍🏽 I agree that it looks more intuitive that way. I'll update this later today.

g1eny0ung commented:
> Hi @g1eny0ung I found some installed themes not in the `select` list no matter which mode selected. Could you try this theme plugin?
> 
> <img alt="CleanShot 2022-05-17 at 09 39 01@2x" width="409" src="https://user-images.githubusercontent.com/1779837/168710629-89c253f4-b848-4256-9635-a56c3febffd4.png">
> 

@xyhp915 This is because the paper theme hasn't set the `mode` field in https://github.com/wirtzdan/logseq-paper-theme/blob/master/package.json.

Due to this PR aims to split light and dark themes into different CSS files. Should we force theme authors to be compatible with this change? Or put the themes without `mode` into **Uncategorized** uniformly and then display them on the page.

xyhp915 commented:
> Due to this PR aims to split light and dark themes into different CSS files. Should we force theme authors to be compatible with this change? Or put the themes without mode into Uncategorized uniformly and then display them on the page.

Hi @g1eny0ung I think compatibility with these themes is still necessary. I remember that many themes have been converted from `custom.css`, most of which already contain two theme modes, and when transforming it to a plugin, the author did not explicitly declare `mode` for the convenience of sharing some CSS code type. 
So when designing the plugin system, if `mode` is empty, it means that both modes are supported. I think we could also be compatible in this way, however, we still recommend that developers declare the `mode` type explicitly.

g1eny0ung commented:
> > Due to this PR aims to split light and dark themes into different CSS files. Should we force theme authors to be compatible with this change? Or put the themes without mode into Uncategorized uniformly and then display them on the page.
> 
> Hi @g1eny0ung I think compatibility with these themes is still necessary. I remember that many themes have been converted from `custom.css`, most of which already contain two theme modes, and when transforming it to a plugin, the author did not explicitly declare `mode` for the convenience of sharing some CSS code type. So when designing the plugin system, if `mode` is empty, it means that both modes are supported. I think we could also be compatible in this way, however, we still recommend that developers declare the `mode` type explicitly.

So it's hard to put these `no mode` themes into the select panel like vscode, they will disrupt the themes that already distinguish mode.

If that's the case I would prefer the way it's handled now, choosing different themes in different modes. We can add themes that don't have mode to the selection list.

g1eny0ung commented:
@xyhp915 I've already addressed the above comments in [445d96f](https://github.com/logseq/logseq/pull/4741/commits/445d96f34884a3ad0e1c32a2af783b8f1b71b64d), below is a screenshot:

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/15034155/169068086-3ac3d25d-6162-40be-a15a-2e1b735c8c45.png">

PTAL again. 😃

xyhp915 reviewed (approved):
LGTM.

nelson-dai-17Live commented:
@g1eny0ung 
I've installed the [Gunmetal Theme](https://github.com/blueteafrog/logseq-gunmetal-theme) which should have dark and light styles, but only light style is listed. Could you take some time to look into this?

<img width="760" alt="image" src="https://user-images.githubusercontent.com/90598131/170163368-8038df36-90dd-49d3-b993-09c8b8991cdd.png">

g1eny0ung commented:
> Update: Never mind, dark style shows in the list after I change `Theme modes` to `dark` in Settings.
> 
> @g1eny0ung I've installed the [Gunmetal Theme](https://github.com/blueteafrog/logseq-gunmetal-theme) which should have dark and light styles, but only light style is listed. Could you take some time to look into this?
> 
> <img alt="image" width="760" src="https://user-images.githubusercontent.com/90598131/170163368-8038df36-90dd-49d3-b993-09c8b8991cdd.png">

Yep. This feature splits the `light` and `dark` themes. This means you need to select twice if you want to have both custom light and dark themes.

Selecting light and dark themes in the same panel could be a enhancement that I will try to update sometime in the future.

chiahsien commented:
Not sure if it related to this enhancement.

1. Change to a theme other than `Default` theme.
2. Restart Logseq.
3. Theme switcher highlights `Default` theme instead of current active theme.

![image](https://user-images.githubusercontent.com/1225437/170708084-0aa8a981-8090-444c-94eb-64c954a6a1d6.png)

g1eny0ung commented:
> Not sure if it related to this enhancement.
> 
> 1. Change to a theme other than `Default` theme.
> 2. Restart Logseq.
> 3. Theme switcher highlights `Default` theme instead of current active theme.
> 
> ![image](https://user-images.githubusercontent.com/1225437/170708084-0aa8a981-8090-444c-94eb-64c954a6a1d6.png)

Yes, it's related and this problem has been resolved in https://github.com/logseq/logseq/pull/5437.
